plugins/links/links.7
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD-PLUGIN-LINKS 7
.Os
.\" NAME
.Sh NAME
.Nm links
.Nd irccd links plugin
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
is used to analyze links sent on channels. It will load the web page and extract
its title on the same channel.
.\" INSTALLATION
.Sh INSTALLATION
The plugin
.Nm
is distributed with irccd. To enable it add it into your
.Xr irccd.conf 5
configuration file.
.Pp
.Bd -literal -offset indent
plugin "links"
.Ed
.\" USAGE
.Sh USAGE
The plugin will automatically fetch web page titles on message that contains
either http://something or https://something.
.Pp
Example of possible output:
.Bd -literal -offset Ds
markand: http://example.org
irccd: Example Domain
.Ed
.\" TEMPLATES
.Sh TEMPLATES
The plugin
.Nm
supports the following template properties:
.Bl -tag -width info
.It Va info
Message to be written when title has been retrieved successfully. Keywords:
.Em channel , nickname , origin , server , title .
.El
.\" BUGS
.Sh BUGS
As HTML pages are not always correct and not openable with a XML parser,
.Nm
actually searches for a <title></title> tag using a regular expression. While
this is not ideal, it usually works in most situations.
.Pp
Only a few set of HTML entity characters are converted to the human readable
format.
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd 1 ,
.Xr irccd-templates 7


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


plugins/links/links.c
/*
 * links.c -- links plugin
 *
 * Copyright (c) 2013-2022 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <curl/curl.h>

#include <irccd/config.h>
#include <irccd/irccd.h>
#include <irccd/limits.h>
#include <irccd/log.h>
#include <irccd/server.h>
#include <irccd/subst.h>
#include <irccd/util.h>

/*
 * Since most websites are now bloated, we need a very large page size to
 * analyse. Use 5MB for now.
 */
#define PAGE_MAX 5242880

struct req {
	CURL *curl;
	CURLM *multi;
	struct irc_server *server;
	struct irc_pollable pollable;
	int status;
	char *link;
	char *chan;
	char *nickname;
	char *origin;
	FILE *fp;
	char buf[];
};

enum {
	TPL_INFO
};

static unsigned long timeout = 3;

static char templates[][512] = {
	[TPL_INFO] = "link from #{nickname}: #{title}"
};

static const struct {
	const char *key;
	int repl;
} entities[] = {
	{ "quot",       '"'     },
	{ "amp",        '&'     },
	{ "apos",       '\''    },
	{ "lt",         '<'     },
	{ "gt",         '>'     },
	{ NULL,         0       }
};

static size_t
callback(char *ptr, size_t size, size_t nmemb, struct req *req)
{
	fprintf(req->fp, "%.*s", (int)(size * nmemb), ptr);

	if (feof(req->fp) || ferror(req->fp))
		return 0;

	return size * nmemb;
}

static char *
parse(struct req *req)
{
	regex_t regex;
	regmatch_t match[2];
	char *ret = NULL;

	if (regcomp(&regex, "<title>([^<]+)<\\/title>", REG_EXTENDED | REG_ICASE) != 0)
		return NULL;

	if (regexec(&regex, req->buf, 2, match, 0) == 0) {
		ret = &req->buf[match[1].rm_so];
		ret[match[1].rm_eo - match[1].rm_so] = '\0';
	}

	regfree(&regex);

	return ret;
}

static int
find_entity(const char *key)
{
	for (size_t i = 0; entities[i].key; ++i)
		if (strcmp(entities[i].key, key) == 0)
			return entities[i].repl;

	return EOF;
}

static const char *
untitle(char *title)
{
	static char ret[256] = {0};
	char *save;
	int repl;
	FILE *fp = NULL;

	if (!(fp = fmemopen(ret, sizeof (ret) - 1, "w")))
		goto fallback;

	for (char *p = title; *p; ) {
		/* Standard character. */
		if (*p != '&') {
			if (fputc(*p++, fp) == EOF)
				goto fallback;

			continue;
		}

		/* HTML entity. */
		save = ++p;

		while (*p && *p != ';')
			++p;

		/* Found an entity. */
		if (*p == ';') {
			*p++ = '\0';

			if ((repl = find_entity(save)) != EOF)
				fputc(repl, fp);
		}
	}

	fclose(fp);

	return ret;

fallback:
	if (fp)
		fclose(fp);

	return title;
}

static const char *
fmt(const struct req *req, char *title)
{
	static char line[IRC_MESSAGE_LEN];

	struct irc_subst subst = {
		.time = time(NULL),
		.flags = IRC_SUBST_DATE | IRC_SUBST_KEYWORDS | IRC_SUBST_IRC_ATTRS,
		.keywords = (const struct irc_subst_keyword []) {
			{ "channel",    req->chan               },
			{ "nickname",   req->nickname           },
			{ "origin",     req->origin             },
			{ "server",     req->server->name       },
			{ "title",      untitle(title)          }
		},
		.keywordsz = 5
	};

	irc_subst(line, sizeof (line), templates[TPL_INFO], &subst);

	return line;
}

static void
complete(struct req *req)
{
	char *title;

	if ((title = parse(req)))
		irc_server_message(req->server, req->chan, fmt(req, title));
}

static void
req_finish(struct req *req)
{
	if (!req)
		return;

	if (req->server)
		irc_server_decref(req->server);

	if (req->curl) {
		if (req->multi)
			curl_multi_remove_handle(req->multi, req->curl);

		curl_easy_cleanup(req->curl);
	}
	if (req->multi)
		curl_multi_cleanup(req->multi);

	if (req->fp)
		fclose(req->fp);

	free(req->link);
	free(req->chan);
	free(req->nickname);
	free(req->origin);
	free(req);
}

static int
pollable_fd(void *data)
{
	struct req *r = data;
	fd_set in, out, exc;
	int fd = 0;

	FD_ZERO(&in);
	FD_ZERO(&out);
	FD_ZERO(&exc);

	curl_multi_fdset(r->multi, &in, &out, &exc, &fd);

	return fd;
}

static void
pollable_want(int *frecv, int *fsend, void *data)
{
	struct req *r = data;
	fd_set in, out, exc;
	int maxfd = 0;

	FD_ZERO(&in);
	FD_ZERO(&out);
	FD_ZERO(&exc);

	curl_multi_fdset(r->multi, &in, &out, &exc, &maxfd);

	if (FD_ISSET(maxfd, &in))
		*frecv = 1;
	if (FD_ISSET(maxfd, &out))
		*fsend = 1;
}

static int
pollable_sync(int frecv, int fsend, void *data)
{
	(void)frecv;
	(void)fsend;

	CURLMsg *msg;
	struct req *r = data;
	int pending, msgsz;

	/*
	 * CURL does its own job reading/sending without taking action on what
	 * have been found.
	 */
	if (curl_multi_perform(r->multi, &pending) < 0)
		return -1;

	/* We only have one handle so we can just assume 0 means complete. */
	if (pending)
		return 0;

	while ((msg = curl_multi_info_read(r->multi, &msgsz))) {
		if (msg->msg != CURLMSG_DONE)
			continue;

		switch (msg->data.result) {
		case CURLE_OPERATION_TIMEDOUT:
			irc_log_warn("links: %s timed out", r->link);
			break;
		case CURLE_OK:
			complete(r);
			break;
		default:
			break;
		}
	}

	return -1;
}

static void
pollable_finish(void *data)
{
	req_finish(data);
}

static void
req(struct irc_server *server, const char *origin, const char *channel, char *link)
{
	assert(link);

	struct req *req;
	struct irc_server_user user;
	int pending;

	if (!(req = calloc(1, sizeof (*req) + PAGE_MAX + 1)))
		goto enomem;
	if (!(req->curl = curl_easy_init()))
		goto enomem;
	if (!(req->multi = curl_multi_init()))
		goto enomem;
	if (!(req->fp = fmemopen(req->buf, PAGE_MAX, "w")))
		goto enomem;

	irc_server_incref(server);
	irc_server_split(origin, &user);

	req->link = link;
	req->server = server;
	req->chan = irc_util_strdup(channel);
	req->nickname = irc_util_strdup(user.nickname);
	req->origin = irc_util_strdup(origin);

	curl_easy_setopt(req->curl, CURLOPT_URL, link);
	curl_easy_setopt(req->curl, CURLOPT_FOLLOWLOCATION, 1L);
	curl_easy_setopt(req->curl, CURLOPT_WRITEFUNCTION, callback);
	curl_easy_setopt(req->curl, CURLOPT_WRITEDATA, req);
	curl_easy_setopt(req->curl, CURLOPT_TIMEOUT, timeout);
	curl_easy_setopt(req->curl, CURLOPT_NOSIGNAL, 1L);
	curl_multi_add_handle(req->multi, req->curl);

	/*
	 * Try immediately to create a socket, otherwise we would poll for a
	 * long time before trying to fetch data.
	 */
	if (curl_multi_perform(req->multi, &pending) != CURLM_OK) {
		req_finish(req);
		return;
	}

	req->pollable.data = req;
	req->pollable.fd = pollable_fd;
	req->pollable.want = pollable_want;
	req->pollable.sync = pollable_sync;
	req->pollable.finish = pollable_finish;
	irc_bot_pollable_add(&req->pollable);

	return;

enomem:
	free(link);
	req_finish(req);

	errno = ENOMEM;
}

void
links_event(const struct irc_event *ev)
{
	char *loc, *end;

	if (ev->type != IRC_EVENT_MESSAGE)
		return;

	/* Parse link. */
	if (!(loc = strstr(ev->message.message, "http://")) &&
	    !(loc = strstr(ev->message.message, "https://")))
	    return;

	/* Compute end to allocate only what's needed. */
	for (end = loc; *end && !isspace(*end); )
		++end;

	req(ev->server, ev->message.origin, ev->message.channel,
	    irc_util_strndup(loc, end - loc));
}

void
links_set_template(const char *key, const char *value)
{
	if (strcmp(key, "info") == 0)
		irc_util_strlcpy(templates[TPL_INFO], value, sizeof (templates[TPL_INFO]));
}

const char *
links_get_template(const char *key)
{
	if (strcmp(key, "info") == 0)
		return templates[TPL_INFO];

	return NULL;
}

const char * const *
links_get_templates(void)
{
	static const char *keys[] = { "info", NULL };

	return keys;
}

void
links_set_option(const char *key, const char *value)
{
	if (strcmp(key, "timeout") == 0)
		timeout = atol(value);
}

const char *
links_get_option(const char *key)
{
	static char out[32];

	if (strcmp(key, "timeout") == 0) {
		snprintf(out, sizeof (out), "%lu", timeout);
		return out;
	}

	return NULL;
}

const char * const *
links_get_options(void)
{
	static const char *keys[] = { "timeout", NULL };

	return keys;
}

const char *links_description = "Parse links from messages";
const char *links_version = IRCCD_VERSION;
const char *links_license = "ISC";
const char *links_author = "David Demelier <markand@malikania.fr>";
